hardware/arduino/avr/cores/arduino/PluggableUSB.cpp

```diff
@@ -0,0 +1,109 @@
+
+
+/* Copyright (c) 2011, Peter Barrett  
+**  
+** Permission to use, copy, modify, and/or distribute this software for  
+** any purpose with or without fee is hereby granted, provided that the  
+** above copyright notice and this permission notice appear in all copies.  
+** 
+** THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL  
+** WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED  
+** WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR  
+** BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES  
+** OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,  
+** WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,  
+** ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS  
+** SOFTWARE.  
+*/
+
+#include "USBAPI.h"
+#include "PluggableUSB.h"
+
+#if defined(USBCON)	
+#ifdef PLUGGABLE_USB_ENABLED
+
+#define MAX_MODULES	6
+
+static u8 lastIf = CDC_ACM_INTERFACE + CDC_INTERFACE_COUNT;
+static u8 lastEp = CDC_FIRST_ENDPOINT + CDC_ENPOINT_COUNT;
+
+class PUSBListNode {
+public:
+	PUSBListNode *next = NULL;
+	PUSBCallbacks cb;
+};
+
+extern u8 _initEndpoints[];
+
+//PUSBCallbacks cbs[MAX_MODULES];
+static u8 modules_count = 0;
+
+static PUSBListNode* rootNode = NULL;
+static PUSBListNode* lastNode = NULL;
+
+int8_t PUSB_GetInterface(u8* interfaceNum)
+{
+	int8_t ret = 0;
+	PUSBListNode* node = rootNode;
+	for (u8 i=0; i<modules_count; i++) {
+		ret = node->cb.getInterface(interfaceNum);
+		node = node->next;
```

Comment:
Posted the comment in a commit, but still applies here. I am not sure if all those functions make so much sense with the retvalue. Also lastNode is not really used.